quixstreams/core/stream/functions.py

```diff
@@ -120,64 +168,146 @@ class FilterFunction(StreamFunction):
 
-        return functools.update_wrapper(wrapper=wrapper, wrapped=self._func)
+        return wrapper
 
-    def get_executor_expanded(self) -> StreamCallable:
-        def wrapper(value: T, func=self._func) -> T:
-            # Apply the function to each item in expanded value and return the
-            # original list
-            for i in value:
-                func(i)
-            return value
 
-        return functools.update_wrapper(wrapper=wrapper, wrapped=self._func)
+class UpdateWithMetadataFunction(StreamFunction):
+    """
+    Wrap a function into an "Update" function.
+
+    The provided function must accept a value, a key, and a timestamp.
+    The callback is expected to mutate the value or to perform some side effect with it.
+
+    The result of the callback is always ignored, and the original input is passed
+    downstream.
+    """
+
+    def __init__(self, func: UpdateWithMetadataCallback):
+        super().__init__(func)
+
+    def get_executor(self, child_executor: VoidExecutor) -> VoidExecutor:
+        def wrapper(value: Any, key: Any, timestamp: int, func=self.func):
+            # Update a single value and forward it
+            func(value, key, timestamp)
+            child_executor(value, key, timestamp)
+
+        return wrapper
+
+
+class TransformFunction(StreamFunction):
+    """
+    Wrap a function into a "Transform" function.
+
+    The provided callback must accept a value, a key and a timestamp.
+    It's expected to return a new value, new key and new timestamp.
+
+    This function must be used with caution, because it can technically change the
+    key.
+    It's supposed to be used by the library internals and not be a part of the public
+    API.
+
+    The result of the callback will always be passed downstream.
+    """
+
+    def __init__(
+        self,
+        func: Union[TransformCallback, TransformExpandedCallback],
+        expand: bool = False,
+    ):
+        super().__init__(func)
+        self.expand = expand
+
+    def get_executor(self, child_executor: VoidExecutor) -> VoidExecutor:
+        if self.expand:
+
+            def wrapper(
+                value: Any,
+                key: Any,
+                timestamp: int,
+                func: TransformExpandedCallback = self.func,
+            ):
+                result = func(value, key, timestamp)
+                for new_value, new_key, new_timestamp in result:
+                    child_executor(new_value, new_key, new_timestamp)
+
+        else:
+
+            def wrapper(
+                value: Any,
+                key: Any,
+                timestamp: int,
+                func: TransformCallback = self.func,
+            ):
+                # Execute a function on a single value and return its result
+                new_value, new_key, new_timestamp = func(value, key, timestamp)
+                child_executor(new_value, new_key, new_timestamp)
+
+        return wrapper
 
 
 def compose(
```

Comment:
There's a couple things I'll mention here:

First, I wonder if it makes sense to just have this live in the `Stream` class (along with the default `sink` function) since you generally aren't composing these functions on their own.

If so, you could potentially simplify it a bit (I didn't test it thoroughly but with a basic example this equivalently worked):

```
    def compose(
        self,
        sink: Callable[[Any, Any, int], None] = _default_sink,
        allow_filters: bool = True,
        allow_updates: bool = True,
        allow_expands: bool = True,
        allow_transforms: bool = True,
    ) -> VoidExecutor:
        tree = self.tree()
        tree.reverse()
        composed = sink
        for node in tree:
            func = node.func
            # Validate that only allowed functions are passed
            if not allow_updates and isinstance(
                    func, (UpdateFunction, UpdateWithMetadataFunction)
            ):
                raise ValueError("Update functions are not allowed")
            elif not allow_filters and isinstance(
                    func, (FilterFunction, FilterWithMetadataFunction)
            ):
                raise ValueError("Filter functions are not allowed")
            elif not allow_transforms and isinstance(func, TransformFunction):
                raise ValueError("Transform functions are not allowed")
            elif not allow_expands and func.expand:
                raise ValueError("Expand functions are not allowed")

            composed = func.get_executor(composed)

        return composed
```

That said, two things: 

1. I also mentioned to you how we should consider removing the extra functions (like the initializing lambda and having a default sink), which this new version would not work as nicely with.

2. once we start doing actual logic branching, it might need to remain something more generalized/independent of the trees for splitting, but hard to say yet what that will look like.

In either case, still might consider moving it to `Stream`...but in all it's a minor thing that may change later anyway.

Reply:
1. I like moving `compose` to `Stream` and simplifying it. I'll change it. Nice catch.

2. I agree that removing the placeholder functions makes sense overall, but I plan to deal with it in future versions. For example, when implementing the branching.
Changing this part doesn't seem essential for this PR.

Reply:
Agreed, I think that's reasonable :+1: